<a href="https://colab.research.google.com/github/Mehak-Kamran/jupiterNotebook/blob/main/ubit24_nlp_TM.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Text Processing with Spacy

---



In [ ]:
docs = ["Kaggle provides notebooks for python.",
       "Python is an easy language.",
       "Kaggle provides many datasets."]

The list of Stop Word and Punctuations which will be removed.



In [ ]:
import spacy
# Creting List of Stop Words
from spacy.lang.en.stop_words import STOP_WORDS
stop_words = spacy.lang.en.stop_words.STOP_WORDS

# Creating list of punctuation marks
import string
punctuations = string.punctuation

print(stop_words)
print("\n===================\n")
print(punctuations)

{'sixty', 'which', 'with', 'why', 'who', 'herein', 'ten', "'re", 'amongst', "n't", '‘d', 'hereupon', 'them', 'so', '’re', 'another', 'therein', "'d", 'whose', 'been', 'just', 'cannot', 'or', '’s', 'now', 'he', 'top', 'not', 'third', 'move', 'no', 'since', 'take', 'front', 'down', 'once', 'i', 'perhaps', 'own', 'anywhere', 'will', 'using', '‘ll', 'between', 'via', 'mine', 'whom', 'where', 'become', 'onto', 'might', 'last', 'at', 'various', 'though', 'did', 'five', 'in', 'for', 'an', 'fifty', 'me', 'whole', "'ll", 'must', 'otherwise', 'whereas', 'anyway', 'n’t', 'often', 'one', 'put', 'everyone', 'itself', 'yours', 'thus', 'very', 'every', 'part', 'while', 'quite', 'should', 'upon', 'through', 'hence', 'yourself', 'doing', 'ca', 'she', 'either', 'those', '‘s', 'such', 'sometimes', 'wherever', 'all', 'twenty', 'elsewhere', 'after', 'first', 'had', 'then', 'else', 'these', 'really', 'mostly', '’ll', 'about', 'by', 'made', 'everything', 'unless', '’m', 'hundred', 'whereafter', 'nor', 'see',

Using SpaCy library to load model of English



In [ ]:
nlp = spacy.load('en_core_web_sm')


Apply the NLP model of English on the given documents. The documents are tokenized and different lexical and syntactical features are assigned to the tokens.

In [ ]:
prc_docs = [nlp(doc) for doc in docs]
print(prc_docs)

#prc_docs = []
#for doc in docs:
  #prc_docs.append(nlp(doc))

for tok in prc_docs[0]:
  print(tok)


[Kaggle provides notebooks for python., Python is an easy language., Kaggle provides many datasets.]
Kaggle
provides
notebooks
for
python
.


Transforming words into their root words.

In [ ]:
print("Before: ", prc_docs)
token_docs = [ [tok.lemma_.lower().strip() for tok in prc_doc] for prc_doc in prc_docs]

#token_docs = []
#for prc_doc in prc_docs:
  #for tok in prc_doc:
    #token_docs.append(tok.lemma_.lower().strip())



print("\nAfter: ", token_docs)

Before:  [Kaggle provides notebooks for python., Python is an easy language., Kaggle provides many datasets.]

After:  [['kaggle', 'provide', 'notebook', 'for', 'python', '.'], ['python', 'be', 'an', 'easy', 'language', '.'], ['kaggle', 'provide', 'many', 'dataset', '.']]


Removing stop words and punctuations.

In [ ]:
print("Before: ", token_docs)

token_docs = [ [tok for tok in token_doc if (tok not in stop_words and tok not in punctuations)] for token_doc in token_docs]
print("\nAfter: ",token_docs)

Before:  [['kaggle', 'provide', 'notebook', 'for', 'python', '.'], ['python', 'be', 'an', 'easy', 'language', '.'], ['kaggle', 'provide', 'many', 'dataset', '.']]

After:  [['kaggle', 'provide', 'notebook', 'python'], ['python', 'easy', 'language'], ['kaggle', 'provide', 'dataset']]


Rebuilding the strings that can be passed to Vectorizer(s).

In [ ]:
s = ''
docs = []
for token_doc in token_docs:
    #for token in token_doc:
    #    s += token + ' '
    #docs.append(s)
    #s = ''
    docs.append(' '.join(token_doc))
print(docs)

['kaggle provide notebook python', 'python easy language', 'kaggle provide dataset']


Transforming the list of tokens to vectors having count of words as the dimensions.

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer

vectorizer = CountVectorizer()
#vectorizer = CountVectorizer(ngram_range=(1,3))
#vectorizer = CountVectorizer(max_features=4)
X = vectorizer.fit_transform(docs)
print("Feature Labels")
print(vectorizer.get_feature_names_out())
print("Sparse Matrix")
print(X)
print("Dense Matrix")
print(X.todense())


Feature Labels
['dataset' 'easy' 'kaggle' 'language' 'notebook' 'provide' 'python']
Sparse Matrix
  (0, 2)	1
  (0, 5)	1
  (0, 4)	1
  (0, 6)	1
  (1, 6)	1
  (1, 1)	1
  (1, 3)	1
  (2, 2)	1
  (2, 5)	1
  (2, 0)	1
Dense Matrix
[[0 0 1 0 1 1 1]
 [0 1 0 1 0 0 1]
 [1 0 1 0 0 1 0]]


# Text Classification

downloading and exploring Brown Corpus

In [ ]:
import nltk
nltk.download('brown')

from nltk.corpus import brown
brown.categories()

[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Unzipping corpora/brown.zip.


['adventure',
 'belles_lettres',
 'editorial',
 'fiction',
 'government',
 'hobbies',
 'humor',
 'learned',
 'lore',
 'mystery',
 'news',
 'religion',
 'reviews',
 'romance',
 'science_fiction']

X and Y will have input and label. Xdocwords will have words/tokens present in each text document.

In [ ]:
Xdocwords = []
Y = []


for category in brown.categories():
  for fileid in brown.fileids(category):
    Xdocwords.append(brown.words(fileid))
    Y.append(category)


Explrong the values in Y and Sdocwords

In [ ]:
print(Y[100])
print(Xdocwords[100][0:20])


belles_lettres
['In', 'a', 'pessimistic', 'assessment', 'of', 'the', 'cold', 'war', ',', 'Eden', 'declared', ':', '``', 'There', 'must', 'be', 'much', 'closer', 'unity', 'within']


In [ ]:
len(Xdocwords)

500

Creating string that will be used by the vectorizer

In [ ]:
Xdoc = []
str = ""
for doc in Xdocwords:
  str = ""
  for w in doc:

    str += w + " "
  Xdoc.append(str)

Using the vectorizer to convert word lists to numeric vectors

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
matrix = CountVectorizer(max_features=200)
X = matrix.fit_transform(Xdoc).toarray()

In [ ]:
print(X)

[[   5    0    1 ...    1   34    4]
 [   6    0    2 ...    6   67    8]
 [   7    0    2 ...    6   85    8]
 ...
 [1811 1005 1072 ...  954 3609  925]
 [1814 1005 1073 ...  956 3621  925]
 [1817 1005 1077 ...  960 3634  925]]


Spliting the dataset into training and test sets.

In [ ]:
# split train and test data
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, Y)

In [ ]:
print(len(X_train))
print(len(X_test))

375
125


Using  Naive Bayes calssifier to create the model

In [ ]:
# Naive Bayes
from sklearn.naive_bayes import MultinomialNB
classifier = MultinomialNB()


classifier.fit(X_train, y_train)

# predict class
y_pred = classifier.predict(X_test)


print(y_test[5:15])
print(y_pred[5:15])


['belles_lettres', 'hobbies', 'news', 'news', 'lore', 'mystery', 'hobbies', 'belles_lettres', 'hobbies', 'adventure']
['editorial' 'hobbies' 'news' 'news' 'lore' 'mystery' 'hobbies'
 'belles_lettres' 'hobbies' 'adventure']


Analyzing the prediction made by the classifier

In [ ]:
# Confusion matrix
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score
cm = confusion_matrix(y_test, y_pred)
cr = classification_report(y_test, y_pred)

accuracy = accuracy_score(y_test, y_pred)

/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.7/dist-packages/sklearn/metrics/_classification.py:1318: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [ ]:
print(cr)
print(cm)

                 precision    recall  f1-score   support

      adventure       0.70      1.00      0.82         7
 belles_lettres       1.00      0.70      0.82        23
      editorial       0.60      1.00      0.75         9
        fiction       1.00      0.60      0.75         5
     government       1.00      0.50      0.67         6
        hobbies       0.70      0.70      0.70        10
          humor       0.14      1.00      0.25         1
        learned       1.00      0.74      0.85        23
           lore       0.79      1.00      0.88        11
        mystery       1.00      1.00      1.00         5
           news       1.00      0.83      0.91        12
       religion       0.60      1.00      0.75         3
        reviews       1.00      1.00      1.00         5
        romance       1.00      0.80      0.89         5
science_fiction       0.00      0.00      0.00         0

       accuracy                           0.81       125
      macro avg       0.77   

Assignment:
1) Convert all the words to lower-case and then note any change in evaluation results.

2) Remove stop words & punctuations and then note any change in evaluation results.

3) Use Tf-idf vectorizer and then note any change in evaluation results.

4) and apply Support Vector Machine (LinearSVM) classifier for training and then note any change in evaluation results.

In [3]:
import spacy
from spacy.lang.en.stop_words import STOP_WORDS
import string
import nltk
from nltk.corpus import brown
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

# Download and load the Brown Corpus
nltk.download('brown')

# Load the Brown Corpus categories and documents
categories = brown.categories()
Xdocwords = []
Y = []

for category in categories:
    for fileid in brown.fileids(category):
        Xdocwords.append(brown.words(fileid))
        Y.append(category)

# Convert lists of words to strings
Xdoc = []
for doc in Xdocwords:
    Xdoc.append(" ".join(doc))

# Step 1: Convert all words to lowercase
Xdoc_lower = [doc.lower() for doc in Xdoc]

# Use the vectorizer to convert word lists to numeric vectors
matrix = CountVectorizer(max_features=200)
X_lower = matrix.fit_transform(Xdoc_lower).toarray()

# Split train and test data
X_train_lower, X_test_lower, y_train, y_test = train_test_split(X_lower, Y)

# Naive Bayes classifier
classifier = MultinomialNB()

classifier.fit(X_train_lower, y_train)
y_pred_lower = classifier.predict(X_test_lower)

# Evaluate the model
cm_lower = confusion_matrix(y_test, y_pred_lower)
cr_lower = classification_report(y_test, y_pred_lower)
accuracy_lower = accuracy_score(y_test, y_pred_lower)

print("Lowercase Model Evaluation:")
print(cr_lower)
print(cm_lower)
print(f"Accuracy: {accuracy_lower}")

# Step 2: Remove Stop Words and Punctuation
nlp = spacy.load('en_core_web_sm')
stop_words = STOP_WORDS
punctuations = string.punctuation

# Remove stop words and punctuation
Xdoc_clean = []
for doc in Xdoc_lower:
    doc_tokens = nlp(doc)
    clean_doc = " ".join([token.text for token in doc_tokens if token.text not in stop_words and token.text not in punctuations])
    Xdoc_clean.append(clean_doc)

# Use the vectorizer to convert word lists to numeric vectors
X_clean = matrix.fit_transform(Xdoc_clean).toarray()

# Split train and test data
X_train_clean, X_test_clean, y_train, y_test = train_test_split(X_clean, Y)

# Naive Bayes classifier
classifier.fit(X_train_clean, y_train)
y_pred_clean = classifier.predict(X_test_clean)

# Evaluate the model
cm_clean = confusion_matrix(y_test, y_pred_clean)
cr_clean = classification_report(y_test, y_pred_clean)
accuracy_clean = accuracy_score(y_test, y_pred_clean)

print("Cleaned Model Evaluation:")
print(cr_clean)
print(cm_clean)
print(f"Accuracy: {accuracy_clean}")

# Step 3: Use TF-IDF Vectorizer
tfidf_vectorizer = TfidfVectorizer(max_features=200)
X_tfidf = tfidf_vectorizer.fit_transform(Xdoc_clean).toarray()

# Split train and test data
X_train_tfidf, X_test_tfidf, y_train, y_test = train_test_split(X_tfidf, Y)

# Naive Bayes classifier
classifier.fit(X_train_tfidf, y_train)
y_pred_tfidf = classifier.predict(X_test_tfidf)

# Evaluate the model
cm_tfidf = confusion_matrix(y_test, y_pred_tfidf)
cr_tfidf = classification_report(y_test, y_pred_tfidf)
accuracy_tfidf = accuracy_score(y_test, y_pred_tfidf)

print("TF-IDF Model Evaluation:")
print(cr_tfidf)
print(cm_tfidf)
print(f"Accuracy: {accuracy_tfidf}")

# Step 4: Apply Support Vector Machine (LinearSVM) Classifier
svm_classifier = LinearSVC()

svm_classifier.fit(X_train_tfidf, y_train)
y_pred_svm = svm_classifier.predict(X_test_tfidf)

# Evaluate the model
cm_svm = confusion_matrix(y_test, y_pred_svm)
cr_svm = classification_report(y_test, y_pred_svm)
accuracy_svm = accuracy_score(y_test, y_pred_svm)

print("SVM Model Evaluation:")
print(cr_svm)
print(cm_svm)
print(f"Accuracy: {accuracy_svm}")


[nltk_data] Downloading package brown to /root/nltk_data...
[nltk_data]   Package brown is already up-to-date!
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_

Lowercase Model Evaluation:
                 precision    recall  f1-score   support

      adventure       0.12      0.14      0.13         7
 belles_lettres       0.57      0.40      0.47        20
      editorial       0.50      0.50      0.50         8
        fiction       0.29      0.25      0.27         8
     government       0.57      0.57      0.57         7
        hobbies       0.67      0.57      0.62         7
          humor       0.12      1.00      0.22         1
        learned       0.77      0.65      0.71        26
           lore       0.67      0.17      0.27        12
        mystery       0.22      0.50      0.31         4
           news       0.67      0.89      0.76         9
       religion       0.33      0.67      0.44         3
        reviews       0.00      0.00      0.00         1
        romance       0.64      0.70      0.67        10
science_fiction       0.00      0.00      0.00         2

       accuracy                           0.50       125
 

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


Cleaned Model Evaluation:
                 precision    recall  f1-score   support

      adventure       0.33      0.62      0.43         8
 belles_lettres       0.55      0.50      0.52        24
      editorial       0.40      0.50      0.44         8
        fiction       0.00      0.00      0.00         7
     government       0.50      0.40      0.44         5
        hobbies       0.50      0.38      0.43         8
          humor       0.00      0.00      0.00         3
        learned       0.62      0.36      0.45        14
           lore       0.18      0.18      0.18        11
        mystery       0.29      0.40      0.33         5
           news       0.75      0.82      0.78        11
       religion       0.75      0.50      0.60         6
        reviews       0.30      0.75      0.43         4
        romance       0.29      0.25      0.27         8
science_fiction       0.00      0.00      0.00         3

       accuracy                           0.42       125
   

/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1471: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/svm/_classes.py:



1. Lowercase Model Evaluation:
    - Accuracy: 0.50
    - Highlights:
        - Best performance in news (F1: 0.76)
        - Poor performance in science_fiction (F1: 0.00)

2. Cleaned Model Evaluation:
    - Accuracy: 0.42
    - Highlights:
        - Best performance in news (F1: 0.78)
        - Poor performance in science_fiction (F1: 0.00)

3. TF-IDF Model Evaluation:
    - Accuracy: 0.32
    - Highlights:
        - Best performance in learned (F1: 0.67)
        - Poor performance in reviews (F1: 0.00)

4. SVM Model Evaluation:
    - Accuracy: 0.34
    - Highlights:
        - Best performance in learned(F1: 0.50)
        - Poor performance in science_fiction (F1: 0.00)
        
Observations:
- All models struggle with science_fiction and humor categories.
- Lowercase Model shows the highest overall accuracy.
- TF-IDF and SVM Models have lower accuracy, suggesting they might not be well-suited for this classification task.